File: leetcode-solutions/python/238. Product of Array Except Self.py
# https://leetcode.com/problems/product-of-array-except-self/

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        products = []
        product_left = 1
        # calculte product of left elements
        for i, n in enumerate(nums):
            products.append(product_left)
            product_left *= nums[i]
        
        product_right = 1
        # multiply left elements product with right element's product
        for i in range(len(nums) - 1, -1, -1):
            products[i] *= product_right
            product_right *= nums[i]
        return products

File: leetcode-solutions/ruby/238. Product of Array Except Self.rb
# https://leetcode.com/problems/product-of-array-except-self/

# @param {Integer[]} nums
# @return {Integer[]}
def product_except_self(nums)
    product = []
    
    current_total_product = 1
    nums.size.times do |i|
        product[i] = current_total_product
        current_total_product *= nums[i] if i < nums.size - 1
    end
    
    current_total_product = 1
    (nums.size - 1).downto(0) do |i|
        product[i] = current_total_product * product[i]
        current_total_product *= nums[i] if i > 0
    end
    
    product
end
